File: main/main_adult.m
% main_adult.m
% ------------
% Applies the MED algorithm to the Adult Gait dataset (3-D marker positions).
%
% Dataset
%   Folder:    data/dataset_Adult_Gait/ (searched recursively for *_pos.csv)
%   that is the processed version with positional trajectories of the 
%   original IMU derived file
%   Structure: each CSV contains a header row, with columns
%              [ignored, X, Y, Z] in metres; sampling rate fixed at 100 Hz.
%
% Requirements
%   pkg/MED_pkg.m must be on the MATLAB path (added automatically below).
%   Run from the repository root or ensure mfilename resolves correctly.
%
% Parameters
%   unit    m | min_D 0.003 m | min_T 0.1 s | min_V 0.01 m/s
%   filter  lp 10 Hz, 4th-order Butterworth
%
% Output
%   output/adult/output_adult_matlab.csv — one row per trial, columns: file,
%   ind, task, trial, then D/V/T/N/Nt/W/R2/P and alpha/K/R2_alpha for each
%   dimension (suffix _all, _x, _y, _z).

clear

repo_root  = fileparts(fileparts(mfilename('fullpath')));
addpath(fullfile(repo_root, 'pkg'));

min_D = 0.003;
min_T = 0.1;
min_V = 0.01;
lp    = 10;
order = 4;

data_dir   = fullfile(repo_root, 'data', 'dataset_Adult_Gait');
output_dir = fullfile(repo_root, 'output', 'adult');
if ~exist(output_dir, 'dir'), mkdir(output_dir); end

var_names = {'file', 'ind', 'task', 'trial'};
var_types = {'int16', 'string', 'string', 'string'};

files        = dir(fullfile(data_dir, '**', '*_pos.csv'));
number_files = length(files);

tab = table('Size', [number_files length(var_types)], ...
    'VariableTypes', var_types, 'VariableNames', var_names);

for i = 1 : number_files

    file_path = [files(i).folder filesep files(i).name];
    parts     = split(files(i).name, '_');

    r          = readmatrix(file_path);
    r          = r(:, 2:4);
    sampleRate = 100;

    output = MED_pkg(r, sampleRate, "m", [min_D, min_T, min_V], [lp, order], ...
        ["scaling", "D", "V", "T", "N", "Nt", "W", "R2", "P", "ME", "timeSeries"]);

    if isempty(output) || isempty(output.D.all)
        continue;
    end

    tab.file(i)  = i;
    tab.ind(i)   = parts(2);
    tab.task(i)  = 'walk';
    tab.trial(i) = parts(3);

    dims       = {'all', 'x', 'y', 'z'};
    suffixes   = {'_all', '_x', '_y', '_z'};
    directVars = {'D', 'V', 'T', 'N', 'Nt', 'W', 'R2', 'P'};
    scaleVars  = {'alpha', 'K', 'R2_alpha'};
    fields     = fieldnames(output.D);

    for k = 1 : length(dims)
        d       = dims{k};
        sfx     = suffixes{k};
        hasData = any(strcmp(fields, d));

        for v = 1 : length(directVars)
            varName     = directVars{v};
            targetField = [varName sfx];
            if hasData
                tab.(targetField)(i) = output.(varName).(d);
            else
                tab.(targetField)(i) = nan;
            end
        end

        for v = 1 : length(scaleVars)
            varName     = scaleVars{v};
            targetField = [varName sfx];
            if hasData
                tab.(targetField)(i) = output.scaling.(varName).(d);
            else
                tab.(targetField)(i) = nan;
            end
        end
    end

    disp(i / number_files);
end

if ~isempty(tab)
    writetable(tab, fullfile(output_dir, 'output_adult_matlab.csv'));
end

File: main/main_adult.py
"""
main_adult.py
-------------
Applies the MED algorithm to the Adult Gait dataset (3-D marker positions).

Dataset
-------
Folder:    data/dataset_Adult_Gait/ (searched recursively for *pos.csv)
Folder:    data/dataset_Adult_Gait/ (searched recursively for *_pos.csv)
           that is the processed version with positional trajectories of the 
           original IMU derived file
Structure: each CSV contains a header row; columns are [ignored, X, Y, Z]
           in metres; sampling rate is fixed at 100 Hz. Filename format:
           <prefix>_<subject>_<trial>_pos.csv.

Requirements
------------
pkg/MED_pkg.py (profile_MED, rescaling_MED, and cost_e removed).

Parameters
----------
unit    m  |  min_D 0.003 m  |  min_T 0.1 s  |  min_V 0.01 m/s
filter  lp 10 Hz, 4th-order Butterworth

Output
------
output/adult/output_adult_python.csv — one row per trial, columns: file,
ind, task, trial, then D/V/T/N/Nt/W/R2/P and alpha/K/R2_alpha for each
dimension (suffix _all, _x, _y, _z).
"""

import os
import sys
import glob
import pandas as pd
import numpy as np

repo_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, os.path.join(repo_root, 'pkg'))
import MED_pkg as m

data_dir   = os.path.join(repo_root, 'data', 'dataset_Adult_Gait')
output_dir = os.path.join(repo_root, 'output', 'adult')
os.makedirs(output_dir, exist_ok=True)

tab   = pd.DataFrame(columns=["file", "ind", "task", "trial"])
files = glob.glob(os.path.join(data_dir, "**", "*pos.csv"), recursive=True)

for i, file_path in enumerate(files):
    file_name = os.path.basename(file_path)
    parts     = file_name.split("_")
    subject   = parts[1]
    trial     = parts[2]
    task      = "walk"

    print(f"Processing: {file_name} | Subject: {subject} | Task: {task} | Trial: {trial}")

    try:
        data = pd.read_csv(file_path, header=0).values
        r    = data[:, 1:4]
    except Exception as e:
        print(f"Error reading {file_name}: {e}")
        continue

    output = m.MED(r, 100, 'm',
                   np.array([0.003, 0.1, 0.01]),
                   np.array([10, 4]),
                   ["scaling", "D", "V", "T", "N", "Nt", "W", "R2", "P", "ME", "timeSeries"])

    dims        = {'all': '_all', 'x': '_x', 'y': '_y', 'z': '_z'}
    direct_vars = ['D', 'V', 'T', 'N', 'Nt', 'W', 'R2', 'P']
    scale_vars  = ['alpha', 'K', 'R2_alpha']
    row_data    = {"file": i, "ind": subject, "task": task, "trial": trial}

    for d, suffix in dims.items():
        if d in output.index:
            for var in direct_vars:
                row_data[f"{var}{suffix}"] = output.loc[d, var]
            scaling_obj = output.loc[d, "scaling"]
            for var in scale_vars:
                try:
                    row_data[f"{var}{suffix}"] = scaling_obj[var]
                except (KeyError, TypeError):
                    row_data[f"{var}{suffix}"] = np.nan
        else:
            for var in direct_vars + scale_vars:
                row_data[f"{var}{suffix}"] = np.nan

    tab = pd.concat([tab, pd.DataFrame([row_data])], ignore_index=True)
    print(i)

tab.to_csv(os.path.join(output_dir, 'output_adult_python.csv'), index=False)
